Fix `FileTaskHandler` only read from default executor

jason810496 opened the pull request:
closes: #45529

Also added `test_file_task_handler_with_multiple_executors` to verify that the correct `get_task_log` method of different executors is called.  

<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

jason810496 committed:
Fix FileTaskHandler only read from default executor

o-nikolas commented on airflow/utils/log/file_task_handler.py:
```diff
@@ -313,10 +313,18 @@ def _render_filename(self, ti: TaskInstance, try_number: int, session=NEW_SESSIO
-    def _executor_get_task_log(self) -> Callable[[TaskInstance, int], tuple[list[str], list[str]]]:
-        """This cached property avoids loading executor repeatedly."""
-        executor = ExecutorLoader.get_default_executor()
+    def _get_executor_get_task_log(
```
You've dropped the cached property of this (which you kind of have to now) but we need to replace it with something else otherwise a new executor is going to get constructed every time this is run. I was just working on this until I went through my email and noticed this one!

Here is a snippet of what I was playing around with:
```python
def _executor_get_task_log(
    self, ti: TaskInstance
) -> Callable[[TaskInstance, int], tuple[list[str], list[str]]]:
    executor = str(ti.executor)
    if not self.executor_instances[executor]:
        # We have not yet loaded this executor, so load it now
        if not executor:
            # The ti is not specifying an executor, so fetch the default
            self.executor_instances[executor] = ExecutorLoader.get_default_executor()
        else:
            self.executor_instances[executor] = ExecutorLoader.load_executor(executor)

    return self.executor_instances[executor].get_task_log

```
The creation of the `executor_instances` in `init` isn't show here but that's obviously required.

Does that make sense?

jason810496 replied:
Indeed, I solved it on the `ExecutorLoader` side yesterday by checking whether the executor was already loaded in the `load_executor` method.  

However, after @nradz shared the solution pointing out that we can cache executors mapping to avoid loading the executor repeatedly, I adopted that approach and resolved the issue on the `FileTaskHandler` side.

o-nikolas replied:
Thanks for the updates!

I don't love the use of `init_executors` because that's going to init _all_ executors, which we will very rarely need. Also the two cached_properties you have now feels like a bit code/complexity than we really need. I much prefer just caching the executors we need as we need them. WDYT?

jason810496 replied:
That is a great point!  
I refactored by caching the necessary executors, but made some adjustments on the `ExecutorLoader` side. Specifically, `_get_executor_names` now needs to be called before using `load_executor`, ensuring that only the `ExecutorName` is loaded without constructing executor instances.

o-nikolas replied:
Yupp, makes sense, I think that should be safe. Thought it might be worth re-running this PR with the full suite of tests.

jason810496 committed:
Add cached_property back to avoid loading executors

jason810496 committed:
Add test for multi-executors scenario

nradz commented on tests/utils/test_log_handlers.py:
```diff
@@ -203,6 +203,66 @@ def task_callable(ti):
 
+    @pytest.mark.parametrize(
+        "executor_name",
+        [
```
I think that It would be interesting to add here _None_ to test the default case.

jason810496 replied:
Sure, just added the test case.

jason810496 committed:
Allow to call load_executor without init_executors

jason810496 committed:
Refactor by caching necessary executors

jason810496 committed:
Refactor test with default executor case

o-nikolas commented on airflow/utils/log/file_task_handler.py:
```diff
@@ -179,6 +179,8 @@ class FileTaskHandler(logging.Handler):
         "Operator inherits from empty operator and thus does not have logs"
     )
+    executor_instances: dict[str, BaseExecutor] = {}
+    default_executor_key = "_default_executor"
```
You could just allow `None` as a key type with an or, but this works as well :+1:

Lee-W replied:
Looks like a constant to me. we probably could capitalize it
```suggestion
    DEFAULT_EXECUTOR_KEY = "_default_executor"
```

jason810496 committed:
Fix side effect from executor_loader

jason810496 committed:
Fix KubernetesExecutor test

- Previous test failure is cuased by cache state of executor_instances
- Should set ti.state = RUNNING after ti.run

Lee-W commented on tests_common/test_utils/executor_loader.py:
```diff
@@ -0,0 +1,34 @@
+    from airflow.executors.executor_utils import ExecutorName
+
+
+def clean_executor_loader():
```
nits: we probably could make it a pytest fixture 🤔

jason810496 committed:
Fix side effect from executor_loader

- The side effect only show up in postgres as backend environment, as
  previous fix only resolve side effect in sqlite as backend environment.
- Also refactor clean_executor_loader as pytest fixture with setup
  teardown

jason810496 committed:
Capitalize default executor key

Lee-W commented on providers/tests/cncf/kubernetes/log_handlers/test_log_handlers.py:
```diff
@@ -74,7 +74,9 @@ def teardown_method(self):
     @pytest.mark.parametrize("state", [TaskInstanceState.RUNNING, TaskInstanceState.SUCCESS])
-    def test__read_for_k8s_executor(self, mock_k8s_get_task_log, create_task_instance, state):
+    def test__read_for_k8s_executor(
+        self, mock_k8s_get_task_log, create_task_instance, state, clean_executor_loader
```
I think we can use `usefixtures` here as it's not used in the test method

https://docs.pytest.org/en/stable/how-to/fixtures.html#use-fixtures-in-classes-and-modules-with-usefixtures

Lee-W commented on tests/executors/test_executor_loader.py:
```diff
@@ -36,23 +35,17 @@ class FakeExecutor:
-    def setup_method(self) -> None:
-        from airflow.executors import executor_loader
+    @pytest.fixture(autouse=True)
+    def setup_method(self, clean_executor_loader) -> None:
```
Should we merge `setup_method` and `teardown_method` as a fixture and make it `autouse` of `TestExecutorLoader`

Lee-W commented on tests_common/pytest_plugin.py:
```diff
@@ -1567,6 +1567,16 @@ def clean_dags_and_dagruns():
+def clean_executor_loader():
+    from tests_common.test_utils.executor_loader import clean_executor_loader_module
+
+    """Clean the executor_loader state, as it stores global variables in the module, causing side effects for some tests."""
```
```suggestion
    """Clean the executor_loader state, as it stores global variables in the module, causing side effects for some tests."""
    
    from tests_common.test_utils.executor_loader import clean_executor_loader_module
```

jason810496 committed:
Refactor clean_executor_loader fixture
